zoomable images feature

TheStu opened the pull request:
This adds image zoom capabilities to image previews. I've added a dependency (https://github.com/BetterTyped/react-zoom-pan-pinch) to handle the actual image manipulation. I added a zoom in, zoom out and reset button to keep things minimal. Additionally you can scroll with the scroll wheel and explore the image with your mouse.

The button functionality is already somewhat tested in their repo, so I haven't added any tests here.

This partially addresses this issue: https://github.com/mattermost/mattermost/issues/15658 though some design work / changes might be needed.

![zoom](https://github.com/mattermost/mattermost/assets/1694325/9a150644-6fc6-4d01-bf5a-fcf734c8093f)

TheStu committed:
zoomable images feature

mattermost-build commented:
Hello @TheStu,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

esethna commented:
Thanks @TheStu! So since we're good on the dependency front I'd suggest re-working the UI to match the designs. I'll add the designer on here if you have any questions. cc// @abhijit-singh

hmhealey commented:
Here's the link again to the design to save a few clicks

https://www.figma.com/file/i86bHWcB68HMiaJNoVxmsq/Image-Preview-Modal?type=design&node-id=125-0&mode=design&t=KGOuz7Dvks2UYJ2s-0

hmhealey commented on webapp/channels/src/components/file_preview_modal/image_preview.tsx:
```diff
@@ -36,13 +37,41 @@ export default function ImagePreview({fileInfo, canDownloadFiles}: Props) {
+                    <>
+                        <div className='image_preview_zoom_actions__actions'>
+                            <button
+                                onClick={() => zoomIn()}
```
For these click handlers, you can actually just pass them in directly as `onClick={zoomIn}` without creating a new function. It mostly just simplifies the code a bit, but I'd assume there's some miniscule performance benefit from not allocating a new function every time this updates

TheStu replied:
There is a type conflict when you pass in the function directly. Maybe there is a way to make the types cooperate? But this was the fix that copilot suggested.

hmhealey replied:
Ah, I see. `zoomIn` takes its own parameters and passing it directly tries to pass the click event as the first one of those. These are fine then

hmhealey commented on webapp/package.json:
```diff
@@ -19,7 +19,8 @@
   "dependencies": {
-    "react-intl": "6.5.5"
+    "react-intl": "6.5.5",
+    "react-zoom-pan-pinch": "3.4.3"
```
Would you mind moving this to `webapp/channels/package.json` instead of here? We try to keep dependencies that are only for the web app there while this contains shared ones

TheStu replied:
done

hmhealey reviewed (changes requested):
The code looks pretty good barring anything that might have to change to fit the designs. There's just a couple small things I'd like changed

mattermost-build commented:
E2E tests not automatically triggered, because the PR is not in a mergeable state. Please update the branch with the base branch and resolve outstanding conflicts.

abhijit-singh reviewed (changes requested):
Thanks @TheStu! This is a great improvement. The designs shared above have some other functionality as well but for this effort, it'll be great if the order of the buttons can be changed and if we can dock them to the top. 

I put together a quick mockup showing that with just the functionality touched in this PR. Here's the figma file for it if you want to take a deeper look - [Figma file](https://www.figma.com/file/i86bHWcB68HMiaJNoVxmsq/Image-Preview-Modal?type=design&node-id=4138-11410&mode=design&t=I7wgfLzYILUFN24o-4)
![image](https://github.com/mattermost/mattermost/assets/14369607/04bd5d60-bea9-4417-ba7d-9b3a223b6da9)

TheStu committed:
figma design changes

TheStu committed:
move toolbar, fix small screen issues, move package

TheStu commented:
I think I've addressed most of what I can with the additional commits, please let me know what you think @hmhealey @abhijit-singh 

For the zoom dropdown I've left out the 'Page Fit' option and the 'Page Width' option. The page fit option seemed to be made for fitting one page of a pdf on the screen, so it didn't seem like it applied here. The page width option seemed difficult to implement given the structure of the modal and seemed like it would be somewhat covered by the fullscreen functionality.

The dedicated hand tool button seemed difficult (maybe impossible?) to implement with the image zoom library, and the built in hand tool functionality seemed good and intuitive, so I didn't try to implement that.

For fullscreen mode, it's a little hard to tell with my screenshots but it uses the built in browser fullscreen mode (F11).

Note also that _tablet.scss had a breakpoint at 960px and the modal header code had a breakpoint at 768px which was causing a design bug between 768px and 960px (which was exacerbated by the addition of the zoom toolbar) so I've addressed that as well with the changes to _tablet.scss.

![zoom_1](https://github.com/mattermost/mattermost/assets/1694325/51e44bc8-d4ec-48b4-b921-36827947fecf)

fullscreen:
![zoom_2](https://github.com/mattermost/mattermost/assets/1694325/db273d8f-50b9-4d86-a4e7-e993dcca4a17)

hmhealey commented on webapp/channels/src/components/file_preview_modal/image_preview.tsx:
```diff
@@ -27,22 +72,157 @@ export default function ImagePreview({fileInfo, canDownloadFiles}: Props) {
+                },
+                ...[50, 75, 100, 125, 150, 200, 300, 400].map((level) => ({
+                    key: `${level}%`,
+                    value: Math.round((imgEl.naturalWidth / imgEl.clientWidth) * level) / 100,
```
I'm a bit surprised to see that these values are based on the image's aspect ratio instead of being constant values like `1` or `100` for 100%. Is there a particular reason you did it this way? It's a bit odd for me to see `1` being the automatic size above instead of that being some special value

TheStu replied:
So yes for the image you would expect a scale of 1 to be the full size image, but for the zoom library a scale of 1 is the image in it's starting scale, where it is filling the container. So there is essentially a mis-match in the scales.

So you either need to let 1 be the scale of the image in it's container and then derive all the other image zoom levels from that (which is what I've done), or you let 1 be the scale of the original image, and convert the scale values going in to and coming out of the library to that scale.

My intuition was that if you convert the image scale to use the library scale then you can do the conversion math up-front (though the option values need to be re-computed when the screen size changes, which is awkward), whereas if you go the other way you need to do some conversion math every time the scale is changed.

What do you think?

hmhealey replied:
Ah, I see. This makes sense to me then. When I saw you had a state for `baseScale` as well, I thought that was being passed to the library to set what `1` means, not that you were holding onto it to do this sort of math later on. Thanks for the explanation!

hmhealey commented on webapp/channels/src/components/file_preview_modal/image_preview.tsx:
```diff
@@ -27,22 +72,157 @@ export default function ImagePreview({fileInfo, canDownloadFiles}: Props) {
-            className='image_preview'
-            href='#'
+        <TransformWrapper
+            onTransformed={(e) => setScale(e.state.scale)}
```
Can you use `useCallback` for this since it's passed into `TransformWrapper` as a prop?

hmhealey commented on webapp/channels/src/components/file_preview_modal/image_preview.tsx:
```diff
@@ -27,22 +72,157 @@ export default function ImagePreview({fileInfo, canDownloadFiles}: Props) {
+                },
+                {
+                    key: 'actual',
+                    value: Math.round((imgEl.naturalWidth / imgEl.clientWidth) * 1000) / 1000, // round to 3 decimal places to prevent a duplicate value conflict with 100%
```
Are you sure that this won't still conflict in certain cases? I know floating point equality in JS is weird, but it seems like these would still be duplicate values for square images. That said, the problem here might be more that we have two duplicate values in the list.

@abhijit-singh Is it intended that we have both "actual size" and "100%" in this list since they should be the same thing?

TheStu replied:
Yes in theory there could be duplicates if the image was exactly the same size as the container div of the image, but it seemed very unlikely so I just left it. I also just wasn't really sure how to account for that, as `1 === 1.0` is true. Setting a custom value for 'Automatic Zoom' like 'auto' or 0 and then adding some conditional logic to change that to a 1 inside `selectZoomLevel` would probably be the simplest workaround. What do you think?

hmhealey replied:
Yeah, using a custom value for that makes sense to me. You'd have to add an if statement or two to `selectZoomLevel` to turn `auto` into `1`, but I think that's a straightforward approach. I'm sure that'll help the next person to work on this code to understand what's going on here

TheStu replied:
So I think this will only work in the case where the user selects 'automatic zoom' from the select menu. If the user uses the mouse wheel or the plus / minus buttons to zoom then the select menu will show the zoom level, not 'automatic zoom'. What do you think?

hmhealey commented on webapp/channels/src/components/file_preview_modal/image_preview.tsx:
```diff
@@ -27,22 +72,157 @@ export default function ImagePreview({fileInfo, canDownloadFiles}: Props) {
+                                value={scale}
+                                onChange={selectZoomLevel}
+                            >
+                                {scaleOptions.map((o) => o.value).includes(scale) ? null : (
```
What is this case for? I don't understand the hidden option here

TheStu replied:
So @abhijit-singh had suggested following the design of this zoom library as closely as possible: https://mozilla.github.io/pdf.js/web/viewer.html

In that library when you click on the plus button or otherwise get to a zoom level that isn't selectable from the dropdown, the level of zoom you're on is still shown as the value for the select box, and that's the functionality that this hidden option replicates.

hmhealey replied:
Ah, I see. I misread the ternary statement, and I didn't realize it was adding an option when the selected value wasn't in the list at all. This makes sense now

hmhealey reviewed (changes requested):
Looks great so far! I have some feedback, and I also noticed that the zoomed image seems to be cropped to its size when at the automatic zoom level

Automatic|Zoomed In
-|-
<img width="1391" alt="Screenshot 2024-03-26 at 5 01 37 PM" src="https://github.com/mattermost/mattermost/assets/3277310/8857de5d-965f-424d-a93b-3676263fd5f6">|<img width="1391" alt="Screenshot 2024-03-26 at 5 01 42 PM" src="https://github.com/mattermost/mattermost/assets/3277310/ea8dca50-18c9-4190-acc1-49fcfa1ae022">
<img width="1728" alt="Screenshot 2024-03-26 at 5 20 05 PM" src="https://github.com/mattermost/mattermost/assets/3277310/c38af58b-0cfe-4666-acac-02104695affa">|<img width="1728" alt="Screenshot 2024-03-26 at 5 20 11 PM" src="https://github.com/mattermost/mattermost/assets/3277310/bc9f8e0b-9c05-40ac-8eef-5d73b256a940">

TheStu commented:
> Looks great so far! I have some feedback, and I also noticed that the zoomed image seems to be cropped to its size when at the automatic zoom level
> 
> Automatic	Zoomed In
> <img alt="Screenshot 2024-03-26 at 5 01 37 PM" width="1391" src="https://private-user-images.githubusercontent.com/3277310/317045033-8857de5d-965f-424d-a93b-3676263fd5f6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1NTE0MTAsIm5iZiI6MTcxMTU1MTExMCwicGF0aCI6Ii8zMjc3MzEwLzMxNzA0NTAzMy04ODU3ZGU1ZC05NjVmLTQyNGQtYTkzYi0zNjc2MjYzZmQ1ZjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjdUMTQ1MTUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDAzYWE2NGRlMGJkN2FlOGUxMWM5YjhmMWFhYWYyYzQ1M2JlM2VhYTY0Y2U3Yjk2ZGIyMTc3OTc1MWQzY2MwOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ZFSZ92UMroAf9pUh0fBaPdMMN-3pN0hqTttPSjf3Uzo">	<img alt="Screenshot 2024-03-26 at 5 01 42 PM" width="1391" src="https://private-user-images.githubusercontent.com/3277310/317045043-ea8dca50-18c9-4190-acc1-49fcfa1ae022.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1NTE0MTAsIm5iZiI6MTcxMTU1MTExMCwicGF0aCI6Ii8zMjc3MzEwLzMxNzA0NTA0My1lYThkY2E1MC0xOGM5LTQxOTAtYWNjMS00OWZjZmExYWUwMjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjdUMTQ1MTUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2Y1ZDc5N2IyN2Y3OWE2MDFmYzA5YmY1ZDYxMTRhM2E5YmY5NDg0MWE3NTI2ZDgzYjRiZmQyYTNlY2UzMjRkYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.WLaf49knnXJk4Il6nrFYAw1_pbaYXgyoMorHi4S8c7s">
> <img alt="Screenshot 2024-03-26 at 5 20 05 PM" width="1728" src="https://private-user-images.githubusercontent.com/3277310/317045093-c38af58b-0cfe-4666-acac-02104695affa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1NTE0MTAsIm5iZiI6MTcxMTU1MTExMCwicGF0aCI6Ii8zMjc3MzEwLzMxNzA0NTA5My1jMzhhZjU4Yi0wY2ZlLTQ2NjYtYWNhYy0wMjEwNDY5NWFmZmEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjdUMTQ1MTUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDgwYzA4MjhkOWNiYWY1MWQzYjhhYTQ4YTc4NTliN2RmNzc5MGZjNzZlZGNmNjEyNTg3NDViZmMyNjNjMTIyOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.kAfgr_BnnQu_8h1BBKUw77A7YwQRrpTom7iJo81B9oA">	<img alt="Screenshot 2024-03-26 at 5 20 11 PM" width="1728" src="https://private-user-images.githubusercontent.com/3277310/317045101-bc9f8e0b-9c05-40ac-8eef-5d73b256a940.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1NTE0MTAsIm5iZiI6MTcxMTU1MTExMCwicGF0aCI6Ii8zMjc3MzEwLzMxNzA0NTEwMS1iYzlmOGUwYi05YzA1LTQwYWMtOGVlZi01ZDczYjI1NmE5NDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjdUMTQ1MTUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjViYjhmYWNiZmI3MDg4ZTk2NTcwYjkwNjI2NGE4MmY3OTFhZmM4MTQ4YmM3MDM2MDdlN2UxZDY0YTkxNTZiYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.FCCEj3W5JbGn9Nv4NF4UN1xes-oL5xxwO9_nFP5Ttcw">

Ah yes, I've been using the image width to figure out all the math but for a portrait image that would create these issues. I'll re-work it.

hmhealey commented:
I wonder if it would be possible to have the zoomable image container always fill the available space but start any larger image with a scale of below 1 to have the same margin appear as does now. We'd have to do some more math to figure out that starting scale, and we'd have to save it to use as the automatic size, but I think that might work

TheStu committed:
zoom small images / handleTransformed callback

hmhealey commented:
> I've set the container to always fill the available space, and that seems to have been enough for the zoom package to work properly with small images, though I also had to redo how the zoom level is calculated. Note that I've also updated the zoom library version to use a PR I made to that library to fix a bug -- [BetterTyped/react-zoom-pan-pinch#468](https://github.com/BetterTyped/react-zoom-pan-pinch/pull/468)

Nice! I like that you made an upstream fix since I feel like we don't make those often enough.

hmhealey reviewed (changes requested):
Playing with this a bit more, I've noticed a few things which seem a bit off:
1. The container now uses more of the available space, but there's still 48px of padding around the image when you zoom in. Would it be possible to still let the image take up that space when zoomed in?

    I tried moving the padding from the `image_preview__content` div onto the `image_preview__image`, and that seems to work better, but the image isn't loading centred when I do that. If that doesn't work, I think it might be possible to remove the padding from the image and just apply it when we calculate the automatic zoom level
2. Clicking in the padding around the image no longer lets you close the image preview. I thought settings `disablePadding` to true might prevent that, but it doesn't.
3. When you're hovering the mouse over the image and its padding, it shows the pointer cursor. I'd expect it to use the regular cursor over the padding and maybe a dragging cursor when you hover over the image, but I'll defer to @abhijit-singh on that

TheStu commented:
re. 1: I'm not too sure what that would look like. If you remove the padding, the image is going to be right up against the edges of the screen, if the image is bigger than the container anyway (maybe you're testing with images that are smaller than the container?). I think that would not look great? Making the padding smaller though might be a good idea. Currently all the different file types in the preview window have padding of 48px which is why I chose that, but it makes sense to have a smaller padding for images where you want them to be maximally large?

re. 2: I didn't even realize that that was a feature. I think making the image container take up all the space must block whatever clickaway listener is in there. I can try and fix that, with z-index maybe or just a new clickaway listener.

re. 3: I'm assuming this is because the <a> element has 100% height and width. Should be an easy fix but I'll wait for @abhijit-singh to weigh in.

hmhealey commented:
1. We'd still want padding when the image is initially opened like it does right now, but if you zoom in further, the image would expand to fill the whole screen. For that, I'm looking at Google Drive as a reference. It also has the same behaviour in 2/3 where clicking outside of the visible image closes the modal

TheStu committed:
restore image click away functionality

TheStu commented:
I've restored the click away functionality. Because the zoom library adds some elements to the DOM, there wasn't a super clean way to do it, but it works.

I took a look at how Google Drive handles zooming. I don't think that functionality would be possible without forking the zoom library, as the bounding box seems to be a fairly integral part of the library.

hmhealey commented:
What if we were to remove the padding from the div and have the initial scale be something below 1 to make the initial image start with a margin? We'd have to do some extra work to keep the previous initial margins, have it only affect larger images, and to make sure that 100% in the dropdown is actually 100%, but I think it could work.

I gave it a quick shot, and while it doesn't do most of the "extra work" I mentioned above, I think this might work? Let me know what you think since I'm happy to discuss more.

https://github.com/mattermost/mattermost/commit/13951d06263d35d729b9e06d3a9001e4151bcb53

hmhealey commented:
Also, nice work on the click outside logic. That works great

TheStu commented:
> What if we were to remove the padding from the div and have the initial scale be something below 1 to make the initial image start with a margin? We'd have to do some extra work to keep the previous initial margins, have it only affect larger images, and to make sure that 100% in the dropdown is actually 100%, but I think it could work.
> 
> I gave it a quick shot, and while it doesn't do most of the "extra work" I mentioned above, I think this might work? Let me know what you think since I'm happy to discuss more.
> 
> [13951d0](https://github.com/mattermost/mattermost/commit/13951d06263d35d729b9e06d3a9001e4151bcb53)

Ah I see, I thought you were more aiming for the zoomed image underneath the modal header like google drive. This looks good!

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

zackyjsd commented:
anything update?Looking forward to this feature

lieut-data commented:
Thank you everyone for the feedback on this proposed solution. There are outstanding changes requested on this pull request that prevent us from merging this solution as-is. Given the age, I'm going to close this pull request for now, but leave the related issues open in case another community member wants to propose another solution.

francislavoie commented:
What outstanding requested changes @lieut-data? The author was the last person to reply here in any substantive way, nobody from Mattermost was actively requesting changes.

lieut-data commented:
Thanks @francislavoie -- @hmhealey & @M-ZubairAhmed, feel free to re-open if I misjudged the state of this old pull request.
